### PR TITLE
plugin/auto: Reload zones every one minute

### PR DESCRIPTION
**Pull request by mrasu**
### 1. Why is this pull request needed and what does it do?
However README says "Default is one minute.", zone data is not reloaded without declaring `reload` option explicitly.
By this change, data is reloaded every one minute by default.

### 2. Which issues (if any) are related?
#2471 

### 3. Which documentation changes (if any) need to be made?
None

### CODE REVIEW

**Comment by stp-ip** [comment]
This fixes part of the problem. Question is, if this should be fixed differently.

I agree it should set a default, but we might want to fix a bit more.

* [x] fix documentation that `reload` is the SOA reload of changed files and `timeout` is the removal and addition timer
* [x] use `timeout` or `reload` to set the other, if not both are set (better default as this seems expected)
* [x] optionally fix it and only rely on reload and deprecate `timeout`

**Comment by mrasu** [comment]
@stp-ip Thanks for feedback!

Following your response, I change behavior of `timeout` and `reload`.
But because I don't know a way to "deprecate" in this project, I'm not sure my change is exactly what you says.
Are belows fine for you?

After my change:

* when `reload` and `timeout` are set -> both use the value of `reload`.
* when only `reload` is set -> both use the value of `reload`.
* when only `timeout` is set -> both use the value of `timeout`.
* when `reload` and `timeout` are not set -> both are one minute.

If this is fine, I will update documentation.

**Comment by stp-ip** [comment]
Paging @miekg as this might be a breaking or deprecation change depending on the decision.

> when reload and timeout are set -> both use the value of reload.

I would say, if both are set, that might be intentional. So overwriting timeout with reload might create unexpected outcomes until it's deprecated.
So first deprecate and then overwrite timeout with reload.

>when only reload is set -> both use the value of reload.

+1

>when only timeout is set -> both use the value of timeout.

+1

> when reload and timeout are not set -> both are one minute

+1

In general I think we should deprecate TIMEOUT.
The general process in my mind:
* document deprecation "reload will be used, if not defined, and it will be removed in a subsequent version"
* print log message that TIMEOUT within plugin/auto is deprecated and RELOAD should be used + a link to the docs
* 1st next release according to deprecation policy (https://github.com/coredns/coredns#deprecation-policy) overwrite TIMEOUT always with RELOAD
* 2nd next release according to deprecation policy remove TIMEOUT and let the config fail, if it's set

**Comment by mrasu** [comment]
@stp-ip done!

File: plugin/auto/README.md [hunk 1]
```diff
@@ -29,9 +29,10 @@ are used.
   used to extract the origin. **ORIGIN_TEMPLATE** will be used as a template for the origin. Strings
   like `{<number>}` are replaced with the respective matches in the file name, e.g. `{1}` is the
   first match, `{2}` is the second. The default is: `db\.(.*)  {1}` i.e. from a file with the
-  name `db.example.com`, the extracted origin will be `example.com`. **TIMEOUT** specifies how often
-  CoreDNS should scan the directory; the default is every 60 seconds. This value is in seconds.
-  The minimum value is 1 second.
+  name `db.example.com`, the extracted origin will be `example.com`. **TIMEOUT** deprecated.
```
Comment (miekg) [inline plugin/auto/README.md]: The sentence starting with **TIMEOUT** looks broken

**Comment by miekg** [comment]
we should just make sure reload = 60 is set and used. Why is this not the case?
We don't need to deprecate anything and add a new option, current set is fine.

**Comment by stp-ip** [comment]
>we should just make sure reload = 60 is set and used. Why is this not the case?
>We don't need to deprecate anything and add a new option, current set is fine.

There are 2 options both with different outcomes.
The idea is to remove TIMEOUT or at least deprecate it. No need to keep a config value, that we basically overwrite in the long term.
My current idea is that, if TIMEOUT is set to something this should still take effect. Then we deprecate and later remote or overwrite the TIMEOUT with the value of RELOAD to make the transition easier.

So the main way to define both reloading files and reloading records is done via RELOAD.

The 60s default was in there, but was changed to nil. Not sure why so? Defaulting to 60s seems fine.

**Comment by stp-ip** [comment]
Correction. The defaulting to 60s is done. It was just moved.

TL;DR:
If TIMEOUT is set, but not RELOAD -> it works
If RELOAD is set, but not TIMEOUT -> it works
If none is set, we default to 60s -> it works
If TIMEOUT and RELOAD is set, we honor the settings -> it works depending on user settings

**Comment by mrasu** [comment]
Actually, my first commit (https://github.com/coredns/coredns/pull/2516/commits/c59bc4aba0a92ffda1fcba91e56d73a479f6cfea) just set 60 seconds as a default.
I deprecate TIMEOUT following with advice.
So either is fine

Honestly, because two values are for almost same purpose, I think it's fair to deprecate TIMEOUT and use the same configuration for watching files too.
But because I'm pretty new to this repository, I cannot judge whether is better.
So I'd like you to decide which is better.

**Comment by miekg** [comment]
ok, I agree with deprecating TIMEOUT, we can offically drop no_reload as well at the same time. Need a release cycle (to announce and then drop) for it to become active. 

What should I write in the release notes?

**Comment by stp-ip** [comment]
> Honestly, because two values are for almost same purpose, I think it's fair to deprecate TIMEOUT and use the same configuration for watching files too.
> But because I'm pretty new to this repository, I cannot judge whether is better.
> So I'd like you to decide which is better.

I think we agree that the purpose is similar, but changing it how it might be more logical might still break people. This is a change in how it works today so having this backwards compatible until it's deprecated seems better.

I'll test this again this week.

**Comment by stp-ip** [comment]
> ok, I agree with deprecating TIMEOUT, we can offically drop no_reload as well at the same time. Need a release cycle (to announce and then drop) for it to become active.
> 
> What should I write in the release notes?

Suggestion:
`Deprecation: plugin/auto deprecates TIMEOUT and NO_RELOAD and recommends the use of RELOAD. This brings plugin/file and plugin/auto to use the same syntax for reload intervals.`

* [x] test current implementation on backwards compatibility and deprecation (me)
* [x] implementation
* [x] docs
* [x] release notes
* [x] create issue for next release to remove deprecated code (including NO_RELOAD from plugin/file)

**Comment by mrasu** [comment]
Sorry for late reply. I'm sick these days(ful?).
I think I cannot contribute until weekend.
please feel free to commit or create new PR if needed.

:+1:

**Comment by stp-ip** [comment]
> Sorry for late reply. I'm sick these days(ful?).
> I think I cannot contribute until weekend.
> please feel free to commit or create new PR if needed.

Get well. I'll take the time to test this PR and then we can iterate next week.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/auto: ..." ]
>> ok, I agree with deprecating TIMEOUT, we can offically drop no_reload as well at the same time. Need a release cycle (to announce and then drop) for it to become active.
>>
>> What should I write in the release notes?
>
>Suggestion:
>`Deprecation: plugin/auto deprecates TIMEOUT and NO_RELOAD and recommends the use of RELOAD. This brings plugin/file and plugin/auto to use the same syntax for reload intervals.`

sgtm

File: plugin/auto/README.md (hunk 1 above)
Comment (stp-ip) [reply plugin/auto/README.md]:
```suggestion
  name `db.example.com`, the extracted origin will be `example.com`. **TIMEOUT** is deprecated and will be removed in a subsequent version.
```

File: plugin/auto/README.md [hunk 2]
```diff
@@ -29,9 +29,10 @@ are used.
...
   like `{<number>}` are replaced with the respective matches in the file name, e.g. `{1}` is the
   first match, `{2}` is the second. The default is: `db\.(.*)  {1}` i.e. from a file with the
-  name `db.example.com`, the extracted origin will be `example.com`. **TIMEOUT** specifies how often
-  CoreDNS should scan the directory; the default is every 60 seconds. This value is in seconds.
-  The minimum value is 1 second.
+  name `db.example.com`, the extracted origin will be `example.com`. **TIMEOUT** deprecated. 
+  `reload` will be used, if not defined, and it will be removed in a subsequent version
+  (it specifies how often CoreDNS should scan the directory to watch for file removal and addition;
+  the default is every 60 seconds. This value is in seconds. The minimum value is 1 second.)
 * `reload` interval to perform reload of zone if SOA version changes. Default is one minute.
```
Comment (stp-ip) [inline plugin/auto/README.md]:
```suggestion
 * `reload` interval to perform reloads of zones if SOA version changes and zonefiles. Default is one minute.
```

**Review by stp-ip — CHANGES_REQUESTED** [review]
Docs changes suggested.

**Comment by mrasu** [comment]
@stp-ip Thanks! changed!

**Comment by mrasu** [comment]
@stp-ip ping

**Review by stp-ip — APPROVED** [review]

**Comment by stp-ip** [comment]
Thanks for the ping. This seems to have slipped through my notifications.

/approve
/lgtm
